### PR TITLE
MEM: add SPATIALREFERENCE element to set projection

### PR DESCRIPTION
**Pull request by mdsumner**
Adds a missing 'SPATIALREFERENCE' element to the MEM driver DSN format. 

Still need test cases added.

## Tasklist

 - [x] Add test case(s)
 - [x] Add documentation
 - [x] Review
 - [x] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

File: doc/source/drivers/raster/mem.rst [hunk 1]
```diff
@@ -64,6 +64,15 @@ or
 -  GEOTRANSFORM: Set the affine transformation coefficients. 6 real
    numbers with '/' as separator (optional)
 
+Coordinate reference system
```
Comment (rouault) [inline doc/source/drivers/raster/mem.rst:67]: I would not create a new paragraph for that, but included it in the above list. The mention of the version can be done with something like ``SPATIALREFERENCE: (GDAL >= 3.7) Set the projection...``

File: frmts/mem/memdataset.cpp [hunk 2]
```diff
@@ -1213,6 +1213,19 @@ GDALDataset *MEMDataset::Open(GDALOpenInfo *poOpenInfo)
...
     }
 
+    /* -------------------------------------------------------------------- */
+    /*      Set Projection Information                                      */
+    /* -------------------------------------------------------------------- */
+
+    pszOption = CSLFetchNameValue(papszOptions, "SPATIALREFERENCE");
+    if (pszOption != nullptr)
+    {
+        if (poDS->m_oSRS.SetFromUserInput(pszOption) != OGRERR_NONE)
```
Comment (rouault) [inline frmts/mem/memdataset.cpp:1223]:
```suggestion
        m_oSRS.SetAxisMappingStrategy(OAMS_TRADITIONAL_GIS_ORDER);
        if (poDS->m_oSRS.SetFromUserInput(pszOption) != OGRERR_NONE)
```

**Review by rouault — COMMENTED** [review]
Editing test_mem_2 in autotest/gdrivers/mem.py to test this would be great. I also see the GEOTRANSFORM option isn't tested

File: doc/source/drivers/raster/mem.rst (hunk 1 above)
Comment (mdsumner) [reply doc/source/drivers/raster/mem.rst:67]: excellent, will do - there are also undocumented elements I'll explore and test them also

File: frmts/mem/memdataset.cpp [hunk 3]
```diff
@@ -1213,6 +1213,20 @@ GDALDataset *MEMDataset::Open(GDALOpenInfo *poOpenInfo)
...
     }
 
+    /* -------------------------------------------------------------------- */
+    /*      Set Projection Information                                      */
+    /* -------------------------------------------------------------------- */
+
+    pszOption = CSLFetchNameValue(papszOptions, "SPATIALREFERENCE");
+    if (pszOption != nullptr)
+    {
+        m_oSRS.SetAxisMappingStrategy(OAMS_TRADITIONAL_GIS_ORDER);
```
Comment (rouault) [inline frmts/mem/memdataset.cpp:1223]:
```suggestion
        poDS->m_oSRS.SetAxisMappingStrategy(OAMS_TRADITIONAL_GIS_ORDER);
```

**Comment by mdsumner** [comment]
bare with me on the tests  please :)  I'm not quite there with the workflow running the python, but I've been verifying the usage in R and it's all good so far. 

I'll get my setup working properly for python and make sure these tests are good

File: doc/source/drivers/raster/mem.rst [hunk 4]
```diff
@@ -63,6 +63,11 @@ or
    the next.
 -  GEOTRANSFORM: Set the affine transformation coefficients. 6 real
    numbers with '/' as separator (optional)
+-  SPATIALREFERENCE: (GDAL >= 3.7) Set the projection. The coordinate reference 
+   systems that can be passed are anything supported by the 
+   OGRSpatialReference.SetFromUserInput() as per '-a_srs' in  
+   :ref:`gdal_translate`.
```
Comment (rouault) [inline doc/source/drivers/raster/mem.rst:69]: can you add a test checking the below precision ? (when doing this in Python, the antislash characters will themselves have to be escaped by antislash!)
```suggestion
   :ref:`gdal_translate`. If the passed string includes double-quote characters (typically WKT),
   it should be surrounded by double-quote characters and the double-quote characters inside it
   should be escaped with anti-slash.
   e.g ``SPATIALREFERENCE="GEOGCRS[\"WGS 84\",[... snip ...],ID[\"EPSG\",4326]]"``
```

Comment (rouault) [reply doc/source/drivers/raster/mem.rst:69]: Above suggestion edited to mention also comma characters as one reason to quote

**Comment by mdsumner** [comment]
I think I messed up the #6183 rebase so badly because I'm treating osgeo/gdal as "upstream", and my fork as the "origin". This is how I've been using forks for PRs for ages, and me blithely running the GDAL-guided rebase instructions got it all wrong. 

I've done a test rebase and squash on this PR and I think it's ok and I understand it!   I'll proceed soon and hopefully get it right this time. All my docker, linting, and testing is now working well so I'm noping to be more productive.

**Comment by mdsumner** [comment]
ugh, no I don't get it ... messed it up again

**Pull request closed without merging**